File: src/sentinel/detectors/system_prompt_exfil.py
"""Detects attempts to extract the system prompt.

The system prompt is often treated as confidential: it can contain business
logic, tool descriptions, or embedded data. Extraction is usually the first move
in a longer attack, because knowing the instructions makes them easier to defeat.

The discriminator here is *whose* instructions are being requested, and *what*
is being dumped:

* "What are your instructions?" targets the model. "What are the instructions
  for assembling this desk?" does not. The possessive is the signal.
* "Repeat the words above" dumps prior context. "Repeat the last paragraph of
  the article" dumps user-supplied content. The object of the verb is the signal.
"""

import re

from sentinel.types import Signal

# Possessive reference to the model's own configuration.
_YOUR_CONFIG_RE = re.compile(
    r"\byour\s+(?:\w+\s+){0,2}?"
    r"(?:system\s+)?(?:prompt|instructions?|directives?|rules?|configuration"
    r"|guidelines?|context|programming|training|persona|messages?)\b",
    re.IGNORECASE,
)

# A dump verb whose object is the prior context rather than user content.
# "above the fold" is excluded: it is a web-layout idiom, not an extraction
# attempt, and it appeared as a false positive during evaluation.
_CONTEXT_DUMP_RE = re.compile(
    r"\b(?:repeat|print|output|show|reveal|display|echo|summarize|recite|list)\b"
    r"(?:\s+\w+){0,3}?\s+"
    r"\b(?:words?|text|everything|all|content|message|prompt|lines?)\b"
    r"(?:\s+\w+){0,2}?\s*"
    r"\b(?:above|before|preceding|prior|earlier|so\s+far)\b"
    r"(?!\s+the\s+fold)",
    re.IGNORECASE,
)

# Explicit references to the initial configuration. A bare "system prompt" is
# deliberately NOT matched here: it is ordinary vocabulary for developers
# discussing their own application, and possessive forms are already covered by
# _YOUR_CONFIG_RE above.
_INITIAL_MESSAGE_RE = re.compile(
    r"\b(?:initial|original|first|very\s+first)\s+"
    r"(?:prompt|instructions?|message|configuration|directive)\b",
    re.IGNORECASE,
)


class SystemPromptExfil:
    """Flags attempts to extract system instructions or prior context."""

    name = "SystemPromptExfil"

    def scan(self, text: str) -> list[Signal]:
        if not text:
            return []

        signals: list[Signal] = []
        for pattern, weight in (
            (_YOUR_CONFIG_RE, 0.8),
            (_CONTEXT_DUMP_RE, 0.85),
            (_INITIAL_MESSAGE_RE, 0.7),
        ):
            for match in pattern.finditer(text):
                signals.append(
                    Signal(
                        detector=self.name,
                        weight=weight,
                        evidence=match.group(0),
                        span=match.span(),
                    )
                )
        return signals

File: web/src/lib/sentinel/detectors/systemPromptExfil.ts
/**
 * Detects attempts to extract the system prompt.
 *
 * ─── YOUR TASK ───────────────────────────────────────────────────────────
 * Tests: `tests/detectors/systemPromptExfil.test.ts`
 * Run:   npm test -- systemPromptExfil
 * Python reference: `src/sentinel/detectors/system_prompt_exfil.py`
 *
 * The system prompt is often confidential: business logic, tool descriptions,
 * embedded data. Extraction is usually the first move in a longer attack.
 *
 * Two discriminators carry this one, and the benign cases target both:
 *
 *   1. WHOSE instructions. "What are your instructions?" targets the model.
 *      "What are the instructions for assembling this desk?" does not. The
 *      possessive is the signal.
 *
 *   2. WHAT is dumped. "Repeat the words above" dumps prior context. "Repeat
 *      the last paragraph of the article" dumps user content. The object of
 *      the verb is the signal.
 *
 * Note a bare "system prompt" must NOT match: it is ordinary vocabulary for a
 * developer discussing their own app. That was a real false positive in the
 * Python version, found by the eval harness.
 * ─────────────────────────────────────────────────────────────────────────
 */

import type { Detector, DetectorName, Signal } from "../types";

export class SystemPromptExfil implements Detector {
  readonly name: DetectorName = "SystemPromptExfil";

  scan(text: string): Signal[] {
    if (!text) return [];
    // TODO(jasper): implement. See instructionOverride.ts for the shape.
    return [];
  }
}
